### PR TITLE
Descriptor type matching

### PR DESCRIPTION
**Pull request by chrisforbes**
Addresses #45. Some cleanup still to be done, but posting for feedback.

### CODE REVIEW

File: layers/draw_state.cpp [hunk 1]
```diff
@@ -1709,6 +1798,26 @@ validate_pipeline_shaders(layer_data *my_data, VkDevice dev, PIPELINE_NODE* pPip
                             pass = VK_FALSE;
                         }
                     }
+                    else if (~descriptor_stage_flags & pStage->stage) {
+                        char type_name[1024];
+                        describe_type(type_name, module, it->second.type_id);
+                        if (log_msg(my_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_DEVICE_EXT, /*dev*/0, __LINE__,
+                                    SHADER_CHECKER_DESCRIPTOR_NOT_ACCESSIBLE_FROM_STAGE, "SC",
+                                    "Shader uses descriptor slot %u.%u (used as type `%s`) but descriptor not accessible from stage %u",
```
Comment (tobine) [inline layers/draw_state.cpp]: To clean up error message I'd like to see the last "%u" changed to "%s" and use "string_VkShaderStageFlagBits(pStage->stage)" to print TXT for shader stage.

File: layers/draw_state.cpp [hunk 2]
```diff
@@ -1709,6 +1798,26 @@ validate_pipeline_shaders(layer_data *my_data, VkDevice dev, PIPELINE_NODE* pPip
...
+                                    it->first.first, it->first.second, type_name, pStage->stage)) {
+                            pass = VK_FALSE;
+                        }
+                    }
+                    else if (!descriptor_type_match(my_data, module, it->second.type_id, descriptor_type)) {
+                        char type_name[1024];
+                        describe_type(type_name, module, it->second.type_id);
+                        if (log_msg(my_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_DEVICE_EXT, /*dev*/0, __LINE__,
+                                    SHADER_CHECKER_DESCRIPTOR_TYPE_MISMATCH, "SC",
+                                    "Type mismatch on descriptor slot %u.%u (used as type `%s`) but descriptor of type %u",
```
Comment (tobine) [inline layers/draw_state.cpp]: To clean up error message I'd like to see the last "%u" changed to "%s" and use "string_VkDescriptorType(descriptor_type)" to print TXT for shader stage.

**Comment by tobine** [comment]
Run clang-format on changes.

This is flagging some errors on "tri --validate" and "cube --validate". I suspect they are valid errors. Can you check those out and include fixes to tri and cube to fix those errors as another patch in this series?

File: layers/draw_state.cpp (hunk 1 above)
Comment (chrisforbes) [inline layers/draw_state.cpp]: Yeah, there's a bunch of places I should do this. There's some handrolled plumbing in SC for stage names that can go this way too.

**Comment by chrisforbes** [comment]
OK, will fix tri/cube's complaints. I suspect many other things are a little broken too.

**Comment by chrisforbes** [comment]
@tobine turned out I'd actually done something quite bogus in has_descriptor_binding, which was upsetting the demos.

**Comment by tobine** [comment]
I pulled latest commits and tested w/ tri/cube/smoke validate and they all run cleanly.
Are you ready for this to be merged yet?
I see that error messages aren't updated, but that's not a must-have.
When you're ready for me to merge this, please re-assign back to me.

**Comment by chrisforbes** [comment]
Rebased again.

**Comment by chrisforbes** [comment]
Rebased /again/ past upstream reformatting of all the layers code.

**Comment by tobine** [comment]
cherry-picked and pushed to master.
Note that I updated your commit to remove the defunct "SHADER_CHECKER_FS_MIXED_BROADCAST" case from vk_validate_layer_details.md. Otherwise a documentation error is flagged on run_all_tests.sh.
I also removed clang-format updates from vk_validate_layer_details.md as doc validation parser doesn't currently like multi-line table entries.

**Pull request closed without merging**